### PR TITLE
server,sql: Add new JoinToken struct, generator, feature flag

### PR DESCRIPTION
**Pull request by itsbilal**
This change adds a new JoinToken struct in the server
package for use in TLS auto-join (see RFC #51991).

It also adds a feature flag to hide this feature except
for opt-in use, as well as a generator function to the status
server that can be used by a future SQL statement that would
generate and return join tokens. The feature flag is also
in the SQL package so it can be used by both SQL and server.

Part of #60632.

WIP until:
- [x] Unit tests for `server.JoinToken`

Release justification: All changes are to new code paths, and
are gated behind a feature flag.
Release note: None.

### CODE REVIEW

File: pkg/server/status.go [hunk 1]
```diff
@@ -2542,3 +2545,39 @@ func (s *statusServer) JobStatus(
...
+	r := rand.New(rand.NewSource(timeutil.Now().UnixNano()))
+	jt.SharedSecret = randutil.RandBytes(r, joinTokenSecretLen)
+
+	certLocator := security.MakeCertsLocator(s.cfg.SSLCertsDir)
+	if ok, err := certLocator.HasNodeCert(); err != nil || !ok {
+		return "", errors.New("cannot generate join token on node without CA cert")
+	}
+
+	// TODO(bilal): This is definitely wrong.
+	f, err := os.Open(certLocator.CACertPath())
```
Comment (itsbilal) [inline pkg/server/status.go:2568]: @aaron-crl This likely calls for a "get me the CA cert" function inside `pkg/security, right? Is there an obvious way to get hold of a certificate manager to get this?

**Review by aaron-crl — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61505)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @knz)

___
*[pkg/server/status.go, line 2568 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MUz31jDeMFZ62ByuN8b-r1-2568:-MUzBCgt7IvfcPln0Em1:bcorxz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be180b18af8547328ac5a73f229e94493580d52/pkg/server/status.go#L2568)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

@aaron-crl This likely calls for a "get me the CA cert" function inside `pkg/security, right? Is there an obvious way to get hold of a certificate manager to get this?
</blockquote></details>

I've got the same need in the GRPC endpoint. I'm using the below code to get it there.

Since you're in server, you should have access to `loadCertificateFile`, so something like this (copied from my GRPC code):

```
// RequestCA makes it possible for a node to request the node-to-node CA certificate.
func (s *adminServer) RequestCA(
	ctx context.Context, req *serverpb.CaRequest,
) (*serverpb.CaResponse, error) {
	cl := security.MakeCertsLocator(s.server.cfg.SSLCertsDir)
	caCert, err := loadCertificateFile(cl.CACertPath())
	if err != nil {
		return nil, errors.Wrapf(
			err,
			"failed to read inter-node cert from disk at %q ",
			cl.CACertPath(),
		)
	}

	[...]
}
```

**Review by itsbilal — COMMENTED** [review]
TFTR! Tests added.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-:-MUzSeFp7NH6vpyDNfFP:bt96oe9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/server/status.go, line 2568 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MUz31jDeMFZ62ByuN8b-r1-2568:-MUzShsJCQiO23fjk8VQ:bd3dq2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be180b18af8547328ac5a73f229e94493580d52/pkg/server/status.go#L2568)):*
<details><summary><i>Previously, aaron-crl wrote…</i></summary><blockquote>

I've got the same need in the GRPC endpoint. I'm using the below code to get it there.

Since you're in server, you should have access to `loadCertificateFile`, so something like this (copied from my GRPC code):

```
// RequestCA makes it possible for a node to request the node-to-node CA certificate.
func (s *adminServer) RequestCA(
	ctx context.Context, req *serverpb.CaRequest,
) (*serverpb.CaResponse, error) {
	cl := security.MakeCertsLocator(s.server.cfg.SSLCertsDir)
	caCert, err := loadCertificateFile(cl.CACertPath())
	if err != nil {
		return nil, errors.Wrapf(
			err,
			"failed to read inter-node cert from disk at %q ",
			cl.CACertPath(),
		)
	}

	[...]
}
```
</blockquote></details>

This is done now - thanks!

**Review by aaron-crl — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61505)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @knz)

___
*[pkg/server/join_token_test.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MUzU9nc6QA-7kzjSPBm:-MUzU9nc6QA-7kzjSPBn:bexcfqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bc6c886b743dc9402d448ad3329f9351ff690db/pkg/server/join_token_test.go#L61)):*
> ```Go
> 	caCertFile := path.Join(tempDir, security.CACertFilename())
> 	caCert := []byte("foobar")
> 	require.NoError(t, ioutil.WriteFile(caCertFile, caCert, 0666))
> ```

nit: This might eventually fall afoul of cert load tests that check permissions. Consider 0600.

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-:-MUzWnJh6jW-Tjq8mUaQ:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/server/join_token_test.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MUzU9nc6QA-7kzjSPBm:-MUzWovJBXIrYPXjRDir:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bc6c886b743dc9402d448ad3329f9351ff690db/pkg/server/join_token_test.go#L61)):*
<details><summary><i>Previously, aaron-crl wrote…</i></summary><blockquote>

nit: This might eventually fall afoul of cert load tests that check permissions. Consider 0600.
</blockquote></details>

Done.

**Comment by itsbilal** [comment]
bors r=aaron-crl

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61505)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[pkg/server/join_token.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MVHCEmN4a3Mc_wZr2IG:-MVHCEmO0qwmdYnh69pU:b-oi6m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d738ecd2ac695281a2a7d059f02e278be7df4f01/pkg/server/join_token.go#L81)):*
> ```Go
> //
> // The format of the text (after base64-decoding) is:
> // <tokenID:uuid.Size><sharedSecret:joinTokenSecretLen><fingerprint:variable><crc:4>
> ```

I would have recommended a version byte in front, so that we can handle multiple formats of join tokens. Can you process this in a followup PR?
___
*[pkg/server/status.go, line 2549 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61505#-MVHCXAE1867gEuKWZ-Q:-MVHCXAF1lhdn0Ghheno:bjmupvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d738ecd2ac695281a2a7d059f02e278be7df4f01/pkg/server/status.go#L2549)):*
> ```Go
> // subsystem directly. The response is a base64 marshaled form of the join token
> // that can be shared to new nodes that want to join this cluster.
> func (s *statusServer) GenerateJoinToken(ctx context.Context) (string, error) {
> ```

Can you explain the rationale of making this a method of statusServer? I though the entire join protocol would be part of the admin interface? Or the new API v2 tree?

**Pull request merged**